storage: handle MVCC range keys in all `NewMVCCIterator` callers

erikgrinaker opened the pull request:
**storage: handle inline values in `MVCCIsSpanEmpty`**

`MVCCIsSpanEmpty` uses an `MVCCIncrementalIterator` to handle time
bounds. However, this will error if it encounters any inline values.
This patch instead uses a regular MVCC iterator when time bounds are not
given (the typical case), which correctly handles inline values.

Release note: None

**batcheval: use `MVCCIsSpanEmpty` in `EndTxn`**

In addition to code deduplication, this also respects MVCC range
tombstones.

Release note: None
  
**kvserver: respect MVCC range keys in `optimizePuts`**

If many point writes are submitted, `optimizePuts()` looks for virgin
keyspace and switches to blind writes to amortize seek costs. This did
not check for MVCC range keys, which could lead to faulty conflict
checks and stats updates.

Release note: None
  
**storage: note functions that ignore MVCC range keys**

Release note: None

Touches #87366.

erikgrinaker committed:
batcheval: use `MVCCIsSpanEmpty` in `EndTxn`

In addition to code deduplication, this also respects MVCC range
tombstones.

Release note: None

erikgrinaker committed:
kvserver: respect MVCC range keys in `optimizePuts`

If many point writes are submitted, `optimizePuts()` looks for virgin
keyspace and switches to blind writes to amortize seek costs. This did
not check for MVCC range keys, which could lead to faulty conflict
checks and stats updates.

Release note: None

erikgrinaker committed:
storage: note functions that ignore MVCC range keys

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87987)
<!-- Reviewable:end -->

erikgrinaker committed:
storage: handle inline values in `MVCCIsSpanEmpty`

`MVCCIsSpanEmpty` uses an `MVCCIncrementalIterator` to handle time
bounds. However, this will error if it encounters any inline values.
This patch instead uses a regular MVCC iterator when time bounds are not
given (the typical case), which correctly handles inline values.

Release note: None

tbg reviewed (approved):
> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87987)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @erikgrinaker)

___
*[`pkg/storage/testdata/mvcc_histories/clear_range` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87987#-NCKxMNa9x29Lsr9Ldag:-NCKxMNa9x29Lsr9Ldah:b-4ozcig) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6cc16a7089beb9189105957c1a66e5ec58ae293/pkg/storage/testdata/mvcc_histories/clear_range#L2)):*
> ```
> 
> # Populate some values
> ```

Pushed a commit with comments.


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
CI failures appear unrelated.

bors r=tbg
